### PR TITLE
BUG: Make digitize, bincount, interp support all floating point types

### LINKED ISSUE #9509: digitize fails on float128
I'm trying to digitize `float128` values, but it looks like there are internal values with fixed, `float64` type causing errors (`float64` works fine).

```
> DT = np.float128
> a = np.linspace(0.0, 1.0, 10, dtype=DT)
> b = np.linspace(0.0, 1.0, 4, dtype=DT)
> d = np.digitize(b, a)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-388-24c1eafc263c> in <module>()
----> 1 d = np.digitize(b, a)

TypeError: Cannot cast array data from dtype('float128') to dtype('float64') according to the rule 'safe'
```

### PR DESCRIPTION
**Pull request by Licht-T**
This fixes #9509.

## What this does
1. Get each type of the array passed to the `digitize`, `bincount` or `interp` function.
2. Compare types and select the most suitable type by `select_most_suitable_float_type` function.
3. Determine the `function_name_@DTYPE@` which execute the calculation by the type selected at 2.

### CODE REVIEW

**Comment by charris** [comment]
Could you document what this does?

**Comment by charris** [comment]
Numerous style issues. See `doc/C_STYLE_GUIDE.rst.txt`.

**Comment by Licht-T** [comment]
@charris Fixed the code style.

I also added the description on the top PR comment. Is this enough?

File: numpy/core/src/multiarray/compiled_base.c.src [hunk 1]
```diff
@@ -8,38 +8,74 @@
...
+        if (a == float_type_list[i]) {
+            a_i = i;
+        }
+        if (b == float_type_list[i]) {
+            b_i = i;
+        }
+    }
+
+    return tolower(float_type_list[(a_i > b_i) ? a_i : b_i]);
+}
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/compiled_base.c.src]: Can we do this with `NPY_DOUBLE` instead of `'D'` etc?

Comment (Licht-T) [reply numpy/core/src/multiarray/compiled_base.c.src]: @eric-wieser Thanks for your review. We can do it but it makes code much complex, I think.
In this function, we compare floating point bit-width including complex types. When we use `NPY_@DTYPE@` enum, we cannot handle the n-bit real type and the n/2-bit complex type as same.

#### Case 1. `NPY_@DTYPE@`
These are not sequential between floating point types (e.g. `NPY_HALF`). Also it's not easy to handle `NPY_@DTYPE@` and `NPY_C@DTYPE@` as same.

#### Case 2. `NPY_@DTYPE@LTR`
An static array cannot hold `toupper(NPY_@DTYPE@LTR)`, and there is no half float complex type. So the `float_type_list[]` becomes
```c
static char float_type_list[] = {'H', NPY_CFLOATLTR, NPY_CDOUBLELTR, NPY_CLONGDOUBLELTR};
```

Comment (Licht-T) [reply numpy/core/src/multiarray/compiled_base.c.src]: @eric-wieser I got an idea. Simply not using `toupper`.
```c
static char
select_most_suitable_float_type(char a, char b) {
    int i = 0, a_i = 0, b_i = 0;

    static char float_type_list[] = {NPY_HALFLTR, NPY_FLOATLTR, NPY_DOUBLELTR, NPY_LONGDOUBLELTR};

    a = tolower(a);
    b = tolower(b);

    if (!(NPY_DOUBLELTR <= a && a <= NPY_LONGDOUBLELTR)) {
        a = NPY_DOUBLELTR;
    }
    if (!(NPY_DOUBLELTR <= a && a <= NPY_LONGDOUBLELTR)) {
        b = NPY_DOUBLELTR;
    }

    for (i = 0; i < 4; i++) {
        if (a == float_type_list[i]) {
            a_i = i;
        }
        if (b == float_type_list[i]) {
            b_i = i;
        }
    }

    return float_type_list[(a_i > b_i) ? a_i : b_i];
}
```

Comment (eric-wieser) [reply numpy/core/src/multiarray/compiled_base.c.src]: I also wonder if we can reuse the ufunc type resolution machinery here.

Comment (Licht-T) [reply numpy/core/src/multiarray/compiled_base.c.src]: Your idea sounds interesting. I'll check them out.

**Comment by eric-wieser** [comment]
Coming back after looking at `np.digitize`, I'd still consider it too broken to use in generic code even after this patch - it converts integers to floats, and fails on objects and datetime types.

**Comment by charris** [comment]
I'm going to close this. @Licht-T Reopen if you want to continue work on this. I would suggest splitting it up to deal with each of the enhancements separately. I also seem to recall other work on digitization that may have come after this.

**Pull request closed without merging**